lib/view_component/i18n.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require "yaml"
+require "pathname"
+
+module ViewComponent
+  module I18n
+    # Expects a component to have a file with the same basename and
+    # the .yml extension, holding translations scoped to the component's path.
+    #
+    # A component YML should look like this:
+    #
+    #   en:
+    #     hello: "Hello World!"
+    #
+    # And inside the component will be possible to call
+    #
+    #   t(".hello") # => "Hello World!"
+    #
+    # Internally the scope will be expanded using the template's path.
+    # As an example, the translation above from within `app/components/greeting/component.html.erb`
+    # will result in an internal I18n key like this:
+    #
+    #   en:
+    #     greeting:
+    #       component:
+    #         hello: "Hello World!"
+    #
+    #
+    def self.load(components_root: "#{Rails.root}/app/components", glob: "**/*component.{yml,yaml}")
```

Comment:
I really love this change, but I think it'd be great if we could make this part of the compilation step.

I think that'd let us more tightly couple components to translation files, removing the need to configure translations. I *think* it might also help with template reloading, but that may need more investigation.

Reply:
Building on that thought a bit, it'd be great if we could use the component's file name to determine the translation file name instead of requiring the translation files to end in `component.*`.

e.g. If I have `modal.rb`, I'd expect the translation file to be called `modal.yml`,  or `modal/modal.yml` (or `.yaml` as an extension).

Reply:
The idea of using the compilation step to add paths to I18n is really neat, and `source_location` is perfect to do that. 

Also about reloading I missed an initializer during the extraction from our application, which would be integrated in `engine.rb` (it's modeled after the original I18n reloading code):

```rb
# Reload component translations whenever we change them.
Rails.application.configure do
  config.after_initialize do |app|
    app.config.file_watcher.new [], "#{Rails.root}/app/components" => ".yml" do
      I18n.reload!
    end.tap do |reloader|
      app.reloaders << reloader
      app.reloader.to_run do
        reloader.execute_if_updated { require_unload_lock! }
      end
    end
  end
end
```

With an hypothetical `modal.rb` component it should be already possible to adjust the glob, even to a generic `**/*.y{a,}ml`, to encompass any YAML in the components folder.

---

So, before proceeding with the implementation inside `compile`, I'd like to have your opinion on a couple of considerations:
- I foresee possible headaches for applications not being strict about only using the translations within the component
- We might still need the reloader to properly pickup changes (renames/deletions) 

The code would go more or less like this:

```rb
def compile
  unless I18n.paths.include? component_locale
    I18n.paths << component_locale
    I18n.reload!
  end
  # ...
end
```

Reply:
> I foresee possible headaches for applications not being strict about only using the translations within the component

Can you share a bit more about that concern? What's an example of "not being strict"?

Reply:
> What I mean is that translations are expected to be there globally, but if we load them only when the component is compiled they might not be there in every environment/situation (think background jobs, rake tasks, partials, etc.) unless we ensure components are all eagerly compiled all the time.

Translations being available globally breaks the implicit and explicit encapsulation that view component assumes. For example, if we have two components with the same name they might have an i18n naming collision. I think it's important to handle that edge case.